Next.mk

```diff
@@ -37,32 +39,110 @@ build: $(current_binary) ## Build the func-e binary
+	@# '-bso0 -bsp0' makes it quiet, except errors
+	@# Wildcards in 7zip will skip the directory. We single-quote to ensure they aren't interpreted by the shell.
+	@7z -bso0 -bsp0 a $@ './$(dir $<)/*.exe'
+else  # Otherwise, assume zip is available
+	@zip -qj $@ $(<:.signed=)
+endif
+	@printf "$(ansi_format_bright)" zip "ok"
+
+# msi-arch is a macro so we can detect it based on the file naming convention
+msi-arch = $(if $(findstring amd64,$1),x64,arm64)
+# Default to a dummy version, which is always lower than a real release
+msi_version=$(VERSION:dev=0.0.1)
+
+# This builds the Windows installer (MSI) using platform-dependent WIX commands.
+$(dist_dir)/func-e_$(VERSION)_%.msi: $(build_dir)/func-e_%/func-e.exe.signed
+	@printf "$(ansi_format_dark)" msi "building $@"
+	@mkdir -p $(dir $@)
+ifeq ($(goos),windows)  # Windows 10 etc use https://wixtoolset.org
+	@candle -nologo -arch $(call msi-arch,$@) -dVersion=$(msi_version) -dBin=$(<:.signed=) packaging/msi/func-e.wxs
+	@light -nologo func-e.wixobj -o $@ -spdb
+	@rm func-e.wixobj
+else  # use https://wiki.gnome.org/msitools
+	@wixl -a $(call msi-arch,$@) -D Version=$(msi_version) -D Bin=$(<:.signed=) -o $@ packaging/msi/func-e.wxs
+endif
+	$(call codesign, $@)
+	@printf "$(ansi_format_bright)" msi "ok"
+
+# Archives are tar.gz, except in the case of Windows, which uses zip.
+archives  := $(non_windows_platforms:%=$(dist_dir)/func-e_$(VERSION)_%.tar.gz) $(windows_platforms:%=$(dist_dir)/func-e_$(VERSION)_%.zip)
+packages  := $(windows_platforms:%=$(dist_dir)/func-e_$(VERSION)_%.msi)
+checksums := $(dist_dir)/func-e_$(VERSION)_checksums.txt
+
+# Darwin doesn't have sha256sum. See https://github.com/actions/virtual-environments/issues/90
+sha256sum := $(if $(findstring darwin,$(goos)),shasum -a 256,sha256sum)
+$(checksums): $(archives) $(packages)
+	@printf "$(ansi_format_dark)" sha256sum "generating $@"
+	@$(sha256sum) $^ > $@
+	@printf "$(ansi_format_bright)" sha256sum "ok"
+
+# dist generates the assets that attach to a release
+# Ex. https://github.com/tetratelabs/func-e/releases/tag/v$(VERSION)
+dist: $(archives) $(packages) $(checksums) ## generates release assets
+
+# this makes a marker file ending in .signed to avoid repeatedly calling codesign
+%.signed: %
+	$(call codesign, $<)
+	@echo > $@
+
+# define macros for multi-platform builds. these parse the filename being built
+go-arch = $(if $(findstring amd64,$1),amd64,arm64)
+go-os   = $(if $(findstring .exe,$1),windows,$(if $(findstring linux,$1),linux,darwin))
 define go-build
 	@printf "$(ansi_format_dark)" build "building $1"
-	@CGO_ENABLED=0 GOOS=$(goos) GOARCH=$(goarch) go build \
+	@CGO_ENABLED=0 GOOS=$(call go-os,$1) GOARCH=$(call go-arch,$1) go build \
 		-ldflags "-s -w -X main.version=$(VERSION)" \
-		-o $1 main.go
-	@printf "$(ansi_format_bright)" build "✔"
+		-o $1 $2
+	@printf "$(ansi_format_bright)" build "ok"
+endef
+
+# This requires osslsigncode package (apt or brew) or latest windows release from mtrojnar/osslsigncode
```

Comment:
Sorry, a question: do we generally assume these tools (including the one from Wixtools above) are installed, and it is not necessary to "assert" their existence?

Reply:
generally it will fail out, and one thing is if we did do an assert, we would want to scope it here because otherwise we end up needing osslsigncode in things that don't require it (ex easy to make e2e accidentally depend on things). So, the thinking is that breaking because file-not-found is as good as breaking explicitly on file-not-found and less work. does that make sense?

Reply:
in any case, if you know a portable search-path operation that will work on windows/darwin/linux and not hairy, we can add it.. just not sure it won't cause more distraction than clarity. in general I wish more things were written in go because we can build them on-demand :D

Reply:
Got it! Thanks for the explanation!